packages/firestore/src/core/sync_engine.ts

```diff
@@ -602,6 +609,62 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
     }
   }
 
+  /**
+   * Takes a snapshot of current mutation queue, and register a user callback
+   * which will be called when all those mutations in the snapshot are either
+   * accepted or rejected by the server.
+   */
+  async registerPendingWritesCallback(callback: Deferred<void>): Promise<void> {
```

Comment:
This method doesn't seem to have to be async. You can just return void.

Reply:
Please follow the pattern that we use here: https://github.com/firebase/firebase-js-sdk/pull/2063/files#diff-3bab365c04d7fc7ae24002035afdd01fR614

It makes the interactions in the client simpler if we limit async methods to their smallest possible scope. Right now, registerPendingWritesCallback only has one consumer, but once it has two or more this would really make a difference (and even with just one consumer, you can already remove two Promise.resolve() invocations that the transpiler adds for you.

Reply:
`this.localStore.getHighestUnacknowledgedBatchId()` returns a promise, which demands it to be async.